File: backend/src/main/kotlin/com/mako/model/Player.kt
package com.mako.model

import com.mako.dto.PlayerDto

/**
 * Available actions for a player on their turn.
 * Determines what buttons/options the UI should show.
 */
enum class AvailableActions {
	/** Player cannot act - folded, all-in, or waiting */
	NONE,

	/** No bet to face - can check, bet, or fold */
	CHECK_BET_FOLD,

	/** Facing a bet - must call, raise, or fold */
	CALL_RAISE_FOLD,

	/** Big blind special case - can check or raise after limps */
	CHECK_RAISE_FOLD
}

/**
 * Base interface for all poker players.
 * Both AI and human players implement this interface.
 * 
 * This allows polymorphic handling of players while still
 * differentiating between AI and human-controlled players.
 */
interface PokerPlayer {
	/** Physical seat at the table (0-9) */
	val seatIndex: Int

	/** Position name (BTN, SB, BB, UTG, etc.) */
	var position: String

	/** Current chip stack */
	var stack: Double

	/** Bet placed in current betting round */
	var currentBet: Double

	/** Player's hole cards */
	val holeCards: MutableList<Card>

	/** Whether player has folded this hand */
	var isFolded: Boolean

	/** Whether player is all-in */
	var isAllIn: Boolean

	/** Last action display string */
	var lastAction: String?

	/** Whether this is the hero (human player) */
	val isHero: Boolean

	/** Whether this player can currently take an action */
	fun canAct(): Boolean = !isFolded && !isAllIn && stack > 0

	/** Reset state for new hand */
	fun reset()

	/** Convert to DTO for API response */
	fun toDto(showCards: Boolean): PlayerDto
}

/**
 * Human player controlled by user input.
 * Cards are always shown to the human player.
 */
data class HumanPlayer(
	override val seatIndex: Int,
	override var position: String = "",
	override var stack: Double,
	override var currentBet: Double = 0.0,
	override val holeCards: MutableList<Card> = mutableListOf(),
	override var isFolded: Boolean = false,
	override var isAllIn: Boolean = false,
	override var lastAction: String? = null
) : PokerPlayer {

	override val isHero: Boolean = true

	override fun reset() {
		holeCards.clear()
		currentBet = 0.0
		isFolded = false
		isAllIn = false
		lastAction = null
	}

	override fun toDto(showCards: Boolean): PlayerDto {
		return PlayerDto(
			seatIndex = seatIndex,
			position = position,
			stack = stack,
			holeCards = holeCards.map { it.toDto() },
			lastAction = lastAction,
			isFolded = isFolded,
			isAllIn = isAllIn,
			currentBet = currentBet,
			isHero = true
		)
	}
}

/**
 * AI player with automated decision making.
 * Cards are hidden until showdown.
 */
data class AiPlayer(
	override val seatIndex: Int,
	override var position: String = "",
	override var stack: Double,
	override var currentBet: Double = 0.0,
	override val holeCards: MutableList<Card> = mutableListOf(),
	override var isFolded: Boolean = false,
	override var isAllIn: Boolean = false,
	override var lastAction: String? = null
) : PokerPlayer {

	override val isHero: Boolean = false

	override fun reset() {
		holeCards.clear()
		currentBet = 0.0
		isFolded = false
		isAllIn = false
		lastAction = null
	}

	override fun toDto(showCards: Boolean): PlayerDto {
		return PlayerDto(
			seatIndex = seatIndex,
			position = position,
			stack = stack,
			holeCards = if (showCards) holeCards.map { it.toDto() } else null,
			lastAction = lastAction,
			isFolded = isFolded,
			isAllIn = isAllIn,
			currentBet = currentBet,
			isHero = false
		)
	}
}

File: apps/api/src/domain/player.ts
import type { Card, Player } from '@mako/shared'

/**
 * Creates a new player.
 */
export function createPlayer(params: {
  seatIndex: number
  stack: number
  isHero: boolean
  position?: string
}): Player {
  return {
    seatIndex: params.seatIndex,
    position: params.position ?? '',
    stack: params.stack,
    holeCards: null,
    lastAction: null,
    isFolded: false,
    isAllIn: false,
    currentBet: 0,
    isHero: params.isHero
  }
}

/**
 * Resets player state for a new hand.
 */
export function resetPlayer(player: Player): void {
  player.holeCards = null
  player.currentBet = 0
  player.isFolded = false
  player.isAllIn = false
  player.lastAction = null
}

/**
 * Checks if player can take an action.
 */
export function canAct(player: Player): boolean {
  return !player.isFolded && !player.isAllIn && player.stack > 0
}

/**
 * Deals hole cards to a player.
 */
export function dealCards(player: Player, cards: Card[]): void {
  player.holeCards = cards
}

/**
 * Converts player to DTO format, optionally hiding cards.
 */
export function playerToDto(player: Player, showCards: boolean): Player {
  return {
    ...player,
    holeCards: showCards ? player.holeCards : null
  }
}

